File: old/src/hammingDistance.zig
pub fn hammingDistance(lhs: []const u8, rhs: []const u8) u32 {
    var res: u32 = 0;
    for (lhs, rhs) |a, b| {
        res += @popCount(a ^ b);
    }
    return res;
}

test "hamming distance" {
    const std = @import("std");
    try std.testing.expectEqual(37, hammingDistance("wokka wokka!!!", "this is a test"));
}

File: src/data/hamming_distance.rs
use crate::Data;

impl Data {
    #[must_use]
    pub fn hamming_distance(&self, other: &Self) -> Option<u32> {
        if self.len() == other.len() {
            let res = (self ^ other).iter().map(|b| b.count_ones()).sum();
            Some(res)
        } else {
            None
        }
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn hamming_distance_works() {
        let lhs = Data::from("this is a test".as_bytes());
        let rhs = Data::from("wokka wokka!!!".as_bytes());
        let res = lhs.hamming_distance(&rhs);
        assert_eq!(Some(37), res);
    }
}
